torch/csrc/autograd/variable.cpp

```diff
@@ -181,19 +181,29 @@ void rebase_history(const Variable& self, Edge gradient_edge) {
   }
 
   set_gradient_edge(self, std::move(gradient_edge));
+  if (self.retains_grad()) {
+    auto old_fn = self.grad_fn()->next_edges()[0].function;
+    old_fn->del_pre_hook(materialize_autograd_meta(self)->grad_fn_prehook_key);
```

Comment:
This actually remove all the hooks that were registered via this Tensor, not just the retain_grad one right?

Reply:
Yes all hooks that were registered with the cpp register_hooks API